### PR TITLE
implement pull request #3412 on master branch

### PR DESCRIPTION
**Pull request by beiwei30**
## What is the purpose of the change

implement pull request #3412 on master branch

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java [hunk 1]
```diff
@@ -44,75 +44,85 @@ public int compare(Object o1, Object o2) {
...
-        } else {
-            com.alibaba.dubbo.common.extension.Activate oa2 = o2.getClass().getAnnotation(com.alibaba.dubbo.common.extension.Activate.class);
-            a2Before = oa2.before();
-            a2After = oa2.after();
-            a2Order = oa2.order();
-        }
-        if ((a1Before.length > 0 || a1After.length > 0
-                || a2Before.length > 0 || a2After.length > 0)
-                && inf != null && inf.isAnnotationPresent(SPI.class)) {
+        if ((a1.before.length > 0 || a1.after.length > 0 || a2.before.length > 0 || a2.after.length > 0) && inf != null) {
```
Comment (khanimteyaz) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: can be refactore by extractive to new util method e.g. 
```
if( inf !=null && (isValidActiveInfo(a1) || isValidActiveInfo(a2)) {
...
}
```
```
private boolean isValidActiveInfo(ActiveInfo activeInfo) {
   return activeInfo.before.length > 0 || activeInfo.after.length > 0
}
```

Will increase the readability. What do you say?

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java [hunk 2]
```diff
@@ -44,75 +44,85 @@ public int compare(Object o1, Object o2) {
...
             ExtensionLoader<?> extensionLoader = ExtensionLoader.getExtensionLoader(inf);
-            if (a1Before.length > 0 || a1After.length > 0) {
+            if (a1.before.length > 0 || a1.after.length > 0) {
```
Comment (khanimteyaz) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: can be changed to 
```
if(isValidActiveInfo(a1)) {
...
}
```

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java [hunk 3]
```diff
@@ -44,75 +44,85 @@ public int compare(Object o1, Object o2) {
...
                 }
-                for (String after : a1After) {
+            for (String after : a1.after) {
                     if (after.equals(n2)) {
                         return 1;
                     }
                 }
             }
-            if (a2Before.length > 0 || a2After.length > 0) {
+            if (a2.before.length > 0 || a2.after.length > 0) {
```
Comment (khanimteyaz) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: can be refcatore to use util method

```
if ( isValidActiveInfo(a2)) {
...
}
```

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java [hunk 4]
```diff
@@ -44,75 +44,85 @@ public int compare(Object o1, Object o2) {
...
+
+    private ActivateInfo parseActivate(Class<?> clazz) {
+        ActivateInfo info = new ActivateInfo();
+        if (clazz.isAnnotationPresent(Activate.class)) {
+            Activate activate = clazz.getAnnotation(Activate.class);
+            info.before = activate.before();
+            info.after = activate.after();
+            info.order = activate.order();
+        } else {
+            com.alibaba.dubbo.common.extension.Activate activate = clazz.getAnnotation(
```
Comment (khanimteyaz) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: Looking at package import **com.alibaba.dubbo.common** is it knowing we have given or just by mistake has come?

Comment (beiwei30) [reply dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: it's for backward compatibility.

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java (hunk 1 above)
Comment (beiwei30) [reply dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: done.

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java (hunk 2 above)
Comment (beiwei30) [reply dubbo-common/src/main/java/org/apache/dubbo/common/extension/support/ActivateComparator.java]: done.

**Comment by beiwei30** [comment]
@khanimteyaz pls. check again.

**Comment by khanimteyaz** [comment]
You made it better than what I have suggested. 👍

**Pull request merged**